src/t_hash.c

```
@@ -46,13 +46,14 @@
  * Hash type Expiry API
  *----------------------------------------------------------------------------*/
 
-static vset *hashTypeGetVolatileSet(robj *o) {
+vset *hashTypeGetVolatileSet(robj *o) {
     serverAssert(o->encoding == OBJ_ENCODING_HASHTABLE);
     return (vset *)hashtableMetadata(o->ptr);
 }
 
 bool hashTypeHasVolatileElements(robj *o) {
-    return ((o->encoding == OBJ_ENCODING_HASHTABLE) && !(vsetIsEmpty(hashTypeGetVolatileSet(o))));
+    serverAssert(o->type == OBJ_HASH);
```

maybe an empty object (NULL) should simply return false?

---

The name implies this function operates on hash objects. It shouldn't handle other types, so returning false for, say, a list would be semantically incorrect. Regardless of the behavior, it's better to keep an assertion to catch such misuse early and detect bugs.

---

NULL is a perfectly valid hash object which is empty.

---

Returning false for NULL is a bad idea - it silently hides bugs. This function is meant to return false if the hash object doesn't contain volatile items, not if the pointer itself is null. If you're passing NULL, that's already a misuse; NULL is not a valid hash object that "happens to be empty." This should assert, not fail quietly. IMO.

---

This is not true IMO. Null is a valid object and all the different commands know how to handle it correctly. given said that. you can place your assert here with a comment explaining why you thought it was a good idea so that IF in the future some change would require it to handle the NULL case it would be understandable.

---

@xbasel I'm in alignment with you here.

I think we should assert if `o` is NULL.  NULL is not a valid object.  A caller should not be calling this function with a null pointer.

This function is predicated that the object is a hash.  To me it makes no sense to silently ignore a NULL object - but then assert on a non-null object of the wrong type.